File: probs/11.cpp
#include <vector>;
using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int lo = 0;
        int hi = height.size() - 1;

        int best = -1;

        while (lo < hi) {
            int area = min(height[lo], height[hi]) * (hi - lo);

            best = max(area, best);

            if (height[lo] > height[hi]) {
                hi--;
            } else {
                lo++;
            }
        }

        return best; 
    }
};

File: zig/11.zig
const std = @import("std");

fn solution(height: []const i32) i32 {
    var lo: usize = 0;
    var hi: usize = height.len - 1;

    var result: i32 = 0;

    while (lo < hi) {
        const h = @as(i32, @intCast(hi - lo)) * @min(height[hi], height[lo]);
        result = @max(h, result);

        if (height[lo] <= height[hi]) {
            lo += 1;
        } else {
            hi -= 1;
        }
    }

    return result;
}

const expect = std.testing.expect;

test "test case #1" {
    const heights = [_]i32{ 1, 8, 6, 2, 5, 4, 8, 3, 7 };

    const sol = solution(&heights);

    try expect(sol == 49);
}

test "test case #2" {
    const heights = [_]i32{ 1, 1 };

    const sol = solution(&heights);

    try expect(sol == 1);
}
